### ISSUE TITLE
Turn off the server and manually process the data.

### ISSUE THREAD

**Issue by kennylajara** [parafoxia/analytix#39]
### Summary

The ability to get the OAuth2 URL instead of showing it in the console and a function to process Google's response without using the server.

### Why?

I am creating a web app and showing the OAuth2 URL in the console is not useful for me as the final users don't see the console. Also, in the case of web apps, you are already running a server, and temporarily running the internal server this library use for authorization may cause many complications for web apps. Some examples are:
- Conflicts with the server that is already running.
- The internal server of the library would try to run several times if several users try to authorize at the same time (a complex logic would need to be implemented to use different ports).
- One internal server instance may run and never be used if the user does not complete the authorization process (more complex needs to be added to handle this scenario).

### Ideal implementation

To have a way to not use the internal server and just show an error message if the user is not authorized yet.
Example:

```python
try:
    client = Analytics.with_secrets("secrets.json", use_server=False)
except NotAuthorized as e:
    # Handle the error however you prefer.
    # `e` contains the error message "You need to authorize analytix; to do so, visit this URL: https://[...]"
```

Also, as the server is turned-off, I'll need a way to process the received data. As this is a webapp, I'll need get the query strings and/or headers however my framework allows me to do it and use a function to save the token.
Example:

```python
Analytics.save_token(
    # ...
)
```

### Contribution

- [x] I wish to contribute to this feature.

> I can contribute by answering questions, providing ideas, and testing. But I don't think I can code it by myself as I don't fully understand how the code works.

### Checklist

- [X] I have searched the issue tracker for duplicate issues and couldn't find any. If this is a follow-up issue, I have specified as such.

**Comment by parafoxia** [comment]
I like the sound of this. I never actually considered the possibility anyone would use this in a web app haha, but I guess now is as good a time as any to do that.

I'm not fully understanding exactly how you want to get the token. Would you want analytix to expose the URL to your webapp to process, or potentially try and do what it's currently doing on the server your webapp is running on?

If you want to (I noticed your note in the contribution section), check out the task/v4 branch. It might be easier to understand because v3 still supports some deprecated features, like manual copy/paste authorisation.

**Comment by kennylajara** [comment]
> I like the sound of this. I never actually considered the possibility anyone would use this in a web app haha, but I guess now is as good a time as any to do that.

I'm glad to know you are open to that idea!

> I'm not fully understanding exactly how you want to get the token. Would you want analytix to expose the URL to your webapp to process, or potentially try and do what it's currently doing on the server your webapp is running on?

I don't think analytix should try to expose the URL as every webapp may use different frameworks, proxies, servers, etc... and this may be difficult to handle.

IMO, every webapp may expose its own URL and receive `state` and `scope` query strings (and any header, if needed). Then, the developer should be able to use some analytix's function and pass the received data. Analytix then saves the Token and does all the things it currently does. The web app is only in charge of receiving the data and forwarding it to Analytix. Analytix shouldn't use the server in "web mode".

> If you want to (I noticed your note in the contribution section), check out the task/v4 branch. It might be easier to understand because v3 still supports some deprecated features, like manual copy/paste authorisation.

Honestly, I'm not very excited about the v4. Looks like the tests are not working properly and the examples are outdated. It's easier for me to know if I break something in the v3 and I understand how to use it. In fact. I think I know how to implement the first part in the v3, I just don't see how you are processing Google's response.

**Comment by parafoxia** [comment]
> Honestly, I'm not very excited about the v4. Looks like the tests are not working properly and the examples are outdated. It's easier for me to know if I break something in the v3 and I understand how to use it.

Fair. V4 is still in development, and I haven't gotten around to rewriting the tests yet.

> I just don't see how you are processing Google's response.

This is done in `webserver.py`. When Google sends the data back, it sends the request to the specified `redirect_uri` (which must be in the secrets file). The actual data is appended as query strings, so for example if the redirect URI was `https://localhost`, the data could be something like `https://localhost?code=something`. This code is actually all you need to get, as getting the token from there can be handled by analytix. [This documentation](https://developers.google.com/identity/protocols/oauth2/openid-connect#server-flow) might help make it clearer.

So for your Google Developers project, your website's URL would need to be set as one of the redirect URIs. From there, your app would need access to the secrets (which analytix could potentially pass to it), then you would need to do the entire OAuth flow yourself. Your server would then receive a query string with the data, and analytix would need the code within that. From there, you can just redirect wherever.

I don't really have experience working with webservers -- in fact, analytix's internal server is the *only* experience I have -- so we'll probably need to exchange ideas a lot haha. Is your app open source? That might help things.

> I think I know how to implement the first part in the v3

Feel free to open a draft PR if you have ideas.

**Comment by parafoxia** [comment]
So I've been thinking about this, and wanted to propose an idea.

The [Google Identity documentation] states authorisation happens in 6 steps. analytix only needs the first 4 of these:

1. [Create an anti-forgery state token](https://developers.google.com/identity/protocols/oauth2/openid-connect#createxsrftoken)
2. [Send an authentication request to Google](https://developers.google.com/identity/protocols/oauth2/openid-connect#sendauthrequest)
3. [Confirm the anti-forgery state token](https://developers.google.com/identity/protocols/oauth2/openid-connect#confirmxsrftoken)
4. [Exchange code for access token and ID token](https://developers.google.com/identity/protocols/oauth2/openid-connect#exchangecode)

When it comes to your web app:

Step 1: analytix has the `oauth.create_state` already which can do this, though it's not too hard to make your own. For this reason analytix can be optionally used here, though I do want to update these utils to be a little nicer for the end user.
Steps 2 and 3: You would be in charge of these. These steps are what analytix uses it's internal webserver for, so these are the bits you would do on your own server in your own framework.
Step 4: analytix should handle this. A new method, `client.exchange`, would be created which would take the authentication code obtained in step 2, and do exactly what it currently does to retrieve an access token and then save it to disk.

I'm thinking the `_try_load_tokens` should be made public, and maybe changed slightly so it saves the tokens within that method rather than returning them out, but either way this isn't particularly difficult.

This approach probably has the least friction when it comes to modifying the v3 code, while providing the opportunity to bypass analytix's internal server entirely. Let me know what you think.

**Comment by kennylajara** [comment]
I just saw your message and it's late at night here. I'll try to find some time to review Google's doc tomorrow.

> Is your app open source? That might help things.

It would be a closed-source project, but it is a brand-new project. If it is necessary, I can give you access to it during the first stage if you agree not to republish it, share it or create derivative work.

> This approach probably has the least friction when it comes to modifying the v3 code, while providing the opportunity to bypass analytix's internal server entirely. Let me know what you think.

A priori, it looks good to me.

Another thing, Is that ok for you if I create a series of small PRs instead of just one big PR?

**Comment by kennylajara** [comment]
Cool.
I may start working on this project this weekend.

**Comment by parafoxia** [comment]
So I've had a big ol' think, and I'm not sure, with the amount of things that need to be considered, this is viable in v3. I've been working on an implementation in v4 with the hope of backporting it, but these changes are extensive, and I'm not sure how backportable they are.

If you do have an idea of how to handle everything you need to be able to handle in v3 go for it, but otherwise I'm gonna limit this to the v4.

**Comment by parafoxia** [comment]
Closed fully in 226537aac5aff825ee846a8263738d2a3ba7e9a0.
